File: dot-prompt-ts/src/asyncClient.ts
import { Transport } from "./transport.js";
import { 
  PromptSchema, 
  CompileResult, 
  RenderResult, 
  InjectResult, 
  type DotPromptEvent,
  ResponseContract
} from "./models.js";
import { validateResponse as _validateResponse } from "./utils.js";
import { createEventStream } from "./events.js";

/**
 * Options for creating a DotPromptAsyncClient.
 */
export interface DotPromptClientOptions {
  baseUrl?: string; // default http://localhost:4041
  timeout?: number;
  apiKey?: string;
  maxRetries?: number;
}

/**
 * Async-first client for Interacting with the dot-prompt container API.
 */
export class DotPromptAsyncClient {
  private transport: Transport;

  constructor(options: DotPromptClientOptions = {}) {
    this.transport = new Transport({
      baseUrl: options.baseUrl || process.env.DOTPROMPT_URL,
      apiKey: options.apiKey || process.env.DOTPROMPT_API_KEY,
      timeout: options.timeout !== undefined ? options.timeout : (process.env.DOTPROMPT_TIMEOUT ? parseInt(process.env.DOTPROMPT_TIMEOUT) : 30000),
      maxRetries: options.maxRetries !== undefined ? options.maxRetries : 3,
    });
  }

  /**
   * Lists all available prompts.
   * 
   * @returns Promise<string[]> - A list of prompt names.
   */
  public async listPrompts(): Promise<string[]> {
    return this.transport.request<string[]>("/api/prompts");
  }

  /**
   * Lists all available prompt collections.
   * 
   * @returns Promise<string[]> - A list of collection names.
   */
  public async listCollections(): Promise<string[]> {
    return this.transport.request<string[]>("/api/collections");
  }

  /**
   * Retrieves the schema for a specific prompt.
   * 
   * @param prompt - The name of the prompt.
   * @param major - Optional major version number.
   * @returns Promise<PromptSchema> - The validated prompt schema.
   */
  public async getSchema(prompt: string, major?: number): Promise<PromptSchema> {
    const path = major 
      ? `/api/schema/${prompt}/${major}` 
      : `/api/schema/${prompt}`;
    
    const data = await this.transport.request(path);
    return PromptSchema.parse(data);
  }

  /**
   * Compiles a prompt with the given parameters.
   * 
   * @param prompt - The name of the prompt.
   * @param params - The input data for compilation.
   * @param options - Optional seed and major version filters.
   * @returns Promise<CompileResult> - The compilation result.
   */
  public async compile(
    prompt: string,
    params: Record<string, any>,
    options: { seed?: number; major?: number } = {}
  ): Promise<CompileResult> {
    const data = await this.transport.request("/api/compile", {
      method: "POST",
      body: { 
        prompt: prompt, 
        params, 
        seed: options.seed, 
        major: options.major 
      },
    });
    return CompileResult.parse(data);
  }

  /**
   * Compiles and renders a prompt by injecting runtime data.
   * 
   * @param prompt - The name of the prompt.
   * @param params - The input data for compilation.
   * @param runtime - The runtime data for rendering.
   * @param options - Optional seed and major version filters.
   * @returns Promise<RenderResult> - The rendering result.
   */
  public async render(
    prompt: string,
    params: Record<string, any>,
    runtime: Record<string, any>,
    options: { seed?: number; major?: number } = {}
  ): Promise<RenderResult> {
    const data = await this.transport.request("/api/render", {
      method: "POST",
      body: { 
        prompt: prompt, 
        params, 
        runtime, 
        seed: options.seed, 
        major: options.major 
      },
    });
    return RenderResult.parse(data);
  }

  /**
   * Injects runtime data into a prompt template.
   * 
   * @param template - The template string to inject into.
   * @param runtime - The runtime data for injection.
   * @returns Promise<InjectResult> - The injection result.
   */
  public async inject(
    template: string, 
    runtime: Record<string, any>
  ): Promise<InjectResult> {
    const data = await this.transport.request("/api/inject", {
      method: "POST",
      body: { template, runtime },
    });
    return InjectResult.parse(data);
  }

  /**
   * Validates a response object against a contract.
   * 
   * @param response - The data to validate.
   * @param contract - The contract containing field definitions.
   * @returns boolean - true if valid, throws ValidationError if not.
   */
  public validateResponse(response: unknown, contract: ResponseContract): boolean {
    return _validateResponse(response, contract);
  }

  /**
   * Returns an async generator for receiving SSE events from the server.
   * 
   * @returns AsyncGenerator<DotPromptEvent>
   */
  public events(): AsyncGenerator<DotPromptEvent> {
    const url = `${this.transport.baseUrl}/api/events`;
    return createEventStream(url, this.transport.apiKey);
  }
}

File: clients/python/src/dotprompt/async_client.py
"""Async client for dot-prompt."""

from typing import Any, AsyncIterator

import httpx

from dotprompt._transport import _Transport
from dotprompt.exceptions import (
    APIClientError,
    ConnectionError,
    MissingRequiredParamsError,
    PromptNotFoundError,
    ServerError,
    TimeoutError,
    ValidationError,
)
from dotprompt.models import CompileResult, PromptSchema


class DotPromptAsyncClient:
    """Async-first client for interacting with the dot-prompt container API."""

    def __init__(self, base_url: str = "http://localhost:4000", **kwargs):
        self._transport = _Transport(base_url=base_url, **kwargs)

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()

    async def close(self) -> None:
        """Close the HTTP client."""
        await self._transport.close()

    async def list_prompts(self) -> list[str]:
        """List all available prompts."""
        response = await self._transport.get("/api/prompts")
        return response.get("prompts", [])

    async def list_collections(self) -> list[str]:
        """List all available prompt collections."""
        response = await self._transport.get("/api/collections")
        return response.get("collections", [])

    async def get_schema(self, prompt_name: str) -> PromptSchema:
        """Get the schema for a specific prompt."""
        data = await self._transport.get(f"/api/schema/{prompt_name}")
        return PromptSchema(**data)

    async def compile(
        self,
        prompt: str,
        params: dict[str, Any],
        options: dict[str, Any] | None = None,
    ) -> CompileResult:
        """Compile a prompt with the given parameters."""
        options = options or {}
        body = {
            "prompt": prompt,
            "params": params,
        }
        if "seed" in options:
            body["seed"] = options["seed"]
        if "version" in options:
            body["version"] = options["version"]

        data = await self._transport.post("/api/compile", body)
        return CompileResult(**data)

    async def render(
        self,
        prompt: str,
        params: dict[str, Any],
        runtime: dict[str, Any] | None = None,
        options: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Render a prompt (compile + fill in variables)."""
        options = options or {}
        body = {
            "prompt": prompt,
            "params": params,
            "runtime": runtime or {},
        }
        if "seed" in options:
            body["seed"] = options["seed"]
        if "version" in options:
            body["version"] = options["version"]

        return await self._transport.post("/api/render", body)

    async def inject(
        self,
        template: str,
        runtime: dict[str, str],
    ) -> dict[str, Any]:
        """Inject runtime variables into a template string."""
        return await self._transport.post("/api/inject", {
            "template": template,
            "runtime": runtime,
        })

    async def validate_response(self, response: dict, contract: dict) -> bool:
        """Validate LLM response against a response contract.
        
        Args:
            response: The LLM response to validate.
            contract: The response contract definition.
        
        Returns:
            True if the response matches the contract, False otherwise.
        """
        # Elixir returns contract with "properties" key, not "fields"
        properties = contract.get("properties", {})
        
        for field_name, field_spec in properties.items():
            if field_name not in response:
                return False
            
            expected_type = field_spec.get("type")
            actual_value = response[field_name]
            
            # Check type
            if expected_type and not self._check_type(actual_value, expected_type):
                return False
        
        return True

    def _check_type(self, value: Any, type_str: str) -> bool:
        """Check if a value matches the expected type."""
        type_mapping = {
            "string": str,
            "number": (int, float),
            "integer": int,
            "boolean": bool,
            "array": list,
            "object": dict,
            "null": type(None),
        }
        
        expected_type = type_mapping.get(type_str)
        if expected_type is None:
            return True
        
        if expected_type == type(None):
            return value is None
        
        return isinstance(value, expected_type)

    @staticmethod
    def _type_mapping(type_str: str):
        """Map type string to Python type (for backwards compatibility)."""
        mapping = {
            "string": str,
            "number": (int, float),
            "integer": int,
            "boolean": bool,
            "array": list,
            "object": dict,
        }
        return mapping.get(type_str, str)
